### PR TITLE
Improve return type of compact

### PR DESCRIPTION
**Pull request by BenoitHiller**
#1602

I'm unsure whether this constitutes a breaking change, as the types do represent an external API that people can rely on so if users were depending on the types returning too much then they could have issues after. If that is the case nearly all type changes would be breaking though.

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -1187,7 +1187,7 @@ declare namespace RamdaAdjunct {
          * Creates an array with all falsy values removed.
          * The values false, null, 0, "", undefined, and NaN are falsy.
          */
-        compact<T>(list: T[]): Array<NonNullable<T>>;
+        compact<T>(list: T[]): Array<Exclude<NonNullable<T>, false | '' | 0>>;
```
Comment (char0n) [inline types/index.d.ts:1190]: Aren't we missing `NaN`. `NaN` is not part of `NonNullable`.

**Review by char0n — COMMENTED** [review]
Thanks! We don't consider enhancement to our types as breaking changes, though It may appear for some consumers at so.

File: types/index.d.ts (hunk 1 above)
Comment (BenoitHiller) [reply types/index.d.ts:1190]: Unfortunately I'm pretty sure that they type of `NaN` is just `number` and that there is no more specific way to match it.

Like right now in the typescript APIs it just has the following to define NaN:
```typescript
declare var NaN: number;
```

So it doesn't have its own type. There seems to be an open proposal to add it:
https://github.com/Microsoft/TypeScript/issues/28682

Comment (char0n) [reply types/index.d.ts:1190]: Cool thank for explanation.

**Review by char0n — APPROVED** [review]
Thanks!

**Pull request merged**